pose_prediction/plugin.cpp

```diff
@@ -14,32 +14,38 @@ class pose_prediction_impl : public pose_prediction {
 		, _m_slow_pose{sb->subscribe_latest<pose_type>("slow_pose")}
         , _m_imu_raw{sb->subscribe_latest<imu_raw_type>("imu_raw")}
         , _m_true_pose{sb->subscribe_latest<pose_type>("true_pose")}
+        , _m_ground_truth_offset{sb->subscribe_latest<Eigen::Vector3f>("ground_truth_offset")}
         , _m_vsync_estimate{sb->subscribe_latest<time_type>("vsync_estimate")}
     { }
 
-    // No paramter get_fast_pose() should just predict to the next vsync
-	// However, we don't have vsync estimation yet.
-	// So we will predict to `now()`, as a temporary approximation
+    // No parameter get_fast_pose() predicts to the next vsync if a vsync
+    // estimate is available, and predicts to `now` otherwise
     virtual fast_pose_type get_fast_pose() const override {
 		const time_type *vsync_estimate = _m_vsync_estimate->get_latest_ro();
 
-        if(vsync_estimate == nullptr) {
-		return get_fast_pose(std::chrono::high_resolution_clock::now());
+        if (vsync_estimate == nullptr) {
+            return get_fast_pose(std::chrono::high_resolution_clock::now());
         } else {
             return get_fast_pose(*vsync_estimate);
         }
-	}
+    }
 
-    virtual pose_type get_true_pose() const override {
+	virtual pose_type get_true_pose() const override {
 		const pose_type* pose_ptr = _m_true_pose->get_latest_ro();
-		return correct_pose(
-			pose_ptr ? *pose_ptr : pose_type{
-				.sensor_time = std::chrono::system_clock::now(),
-				.position = Eigen::Vector3f{0, 0, 0},
-				.orientation = Eigen::Quaternionf{1, 0, 0, 0},
-			}
-		);
-    }
+		pose_type offset_pose;
+
+		// Subtract offset if valid pose pointer, otherwise use zero pose
+		if (pose_ptr) {
+			offset_pose = *pose_ptr;
+			offset_pose.position -= *(_m_ground_truth_offset->get_latest_ro());
```

Comment:
```suggestion
			auto offset = _m_ground_truth_offset->get_latest_ro();
			assert(offset);
			offset_pose.position = offset;
```

We should check the pointer returned by switchboard. It will be null if no data has been pushed.

Reply:
Is this necessary given that `_m_ground_truth_offset` would be non-null if `pose_ptr` is non-null? (no relaxed atomics)

Reply:
I agree that it is not necessary, because _the current implementation_ translates each put into a CAS. I still think it's a good idea anyway.